### PR TITLE
Ft implement pagination 338

### PR DESCRIPTION
**Pull request by beblicarl**
#### Pre-Submission Checklist

-   [x] Your pull request targets the `main` branch.
-   [x] Branch name is descriptive and follows the guidelines [here](https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices).
-   [x] All tests pass . Use `git commit --amend` to amend any fixes.

#### Checklist:

-   [x] All tests are passing.
-   [x] Tested changes locally.
-   [x] Improved documentation or readme where necessary.
-   [x] Your commit messages format follows our commit message guideline [here](https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices#commits).

#### What does this PR do?

https://github.com/tbf-challenge/teamwork-be/issues/338

#### How should this be manually tested?

1. Start the server with npm run dev
2. Make a GET request with query params to this endpoint  `http://localhost:3000/feed?isFlagged=true&cursor=1&limit=1`
3. This successfully returns the sample data below

```
{
    "status": "success",
    "data": [
        {
            "title": "mygifff",
            "imageUrl": "https://ibb.co/M9ChjW9",
            "createdOn": "2022-09-21T16:01:59.541Z",
            "gifId": 42,
            "published": null,
            "flagsCount": 2,
            "likesCount": 0,
            "user": {
                "userId": 7,
                "fullName": "Example Example",
                "profilePictureUrl": null,
                "email": "example@yahoo.com"
            }
        }
    ],
    "metadata": {
        "totalCount": 1
    }
}
```
4. The query params are optional hence kindly play around with them to return different results
5. Run the test with `npm run test` to see all tests passing
#### What are the relevant issues?
https://github.com/tbf-challenge/teamwork-be/issues/338

### CODE REVIEW

File: src/services/posts/fetch-posts.js [hunk 1]
```diff
@@ -14,19 +14,21 @@ const fetchPosts = async(isFlagged) => {
...
+	 let whereClause = `WHERE posts."id" > ${cursor}` 
 	 let orderByClause = `ORDER BY posts."createdAt" DESC`
+	 const limitBy = `LIMIT ${limit}`
 	   
 	 if(isFlagged !== undefined){
 		 selectClause+= `, posts."flagsCount"`
 		 
-		whereClause += `
-			WHERE posts."flagsCount" ${isFlagged ? '>': '='} 0`
+		whereClause = ` ${whereClause}
```
Comment (jocrah) [inline src/services/posts/fetch-posts.js:24]: @beblicarl please could you elaborate on why the change from `+=`

File: src/services/posts/fetch-posts.js [hunk 2]
```diff
@@ -14,19 +14,21 @@ const fetchPosts = async(isFlagged) => {
 	 const fromClause = `FROM posts`
 	  
 	 const joinClause = `INNER JOIN users on posts."userId" = users.id`
-	 let whereClause = ''
+	 let whereClause = `WHERE posts."id" > ${cursor}` 
 	 let orderByClause = `ORDER BY posts."createdAt" DESC`
+	 const limitBy = `LIMIT ${limit}`
```
Comment (jocrah) [inline src/services/posts/fetch-posts.js:19]: `limitClause` is fine

File: src/services/posts/fetch-posts.js [hunk 3]
```diff
@@ -1,8 +1,8 @@
 const db = require("../../db")
 
-const fetchPosts = async(isFlagged) => {
+const fetchPosts = async(isFlagged, cursor, limit) => {
```
Comment (jocrah) [inline src/services/posts/fetch-posts.js:3]: i missed something here in our discussion, cursor will not be always available, especially for the first requests. 

the whereClause which has the cursor should only be used if cursor is defined

File: src/services/posts/fetch-posts.js (hunk 1 above)
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:24]: oh it was a personal choice, but I will change it back

File: src/services/posts/fetch-posts.js (hunk 2 above)
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:19]: sure, thanks

File: src/services/posts/fetch-posts.js (hunk 3 above)
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:3]: alright, i will work on that too

File: docs/swagger.yaml [hunk 4]
```diff
@@ -1795,7 +1795,12 @@ paths:
                           type: boolean
                           description: Published or not
                           example: true
-                        
+                  metadata:        
+                    type: object
+                    description: metadata
+                    example:
```
Comment (jocrah) [inline docs/swagger.yaml:1801]: This has type `object` and should have a `properties` field as we have for the others.

File: src/routes/posts.js [hunk 5]
```diff
@@ -36,7 +36,10 @@ const fetchPosts = catchAsync( async(req, res) => {
 				profilePictureUrl : post.profilePictureUrl,
 				email : post.email
 			})})
-		)
+		),
+		metadata: {
+			totalCount: feed.length
```
Comment (jocrah) [inline src/routes/posts.js:41]: So the aim of the `totalCount` is to give an idea of the total number of posts in the db, so that the client can know if there is more documents to fetch, so its not really about counting the number of documents being returned.

It involves having a count query to the database. 

More information on the reasoning is highlighted in the technical and additional considerations of the issue description

File: src/services/posts/fetch-posts.js [hunk 6]
```diff
@@ -1,7 +1,6 @@
 const db = require("../../db")
 
-const fetchPosts = async(isFlagged) => {
-
+const fetchPosts = async({isFlagged, cursor, limit }) => {
```
Comment (jocrah) [inline src/services/posts/fetch-posts.js:3]: We can handle the default at the service level since that's where the main fetch is happening. That could help to prevent having to pass limit of 20 to the function calls in the tests

```suggestion
const DEFAULT_LIMIT = 20

const fetchPosts = async({isFlagged, cursor, limit = DEFAULT_LIMIT}) => {
```

File: src/schema.js [hunk 7]
```diff
@@ -268,7 +268,9 @@ const updateUserSchema = Joi.object({
 
 const fetchPostsSchema = Joi.object({
 	query: {
-		isFlagged :Joi.boolean()
+		isFlagged :Joi.boolean(),
+		cursor : Joi.number(),
+		limit : Joi.number().default(20)
```
Comment (jocrah) [inline src/schema.js:273]: we can let this be without the default in light of the other suggested change

File: docs/swagger.yaml (hunk 4 above)
Comment (beblicarl) [reply docs/swagger.yaml:1801]: yes please

File: src/services/posts/fetch-posts.js (hunk 6 above)
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:3]: yes please

File: src/schema.js (hunk 7 above)
Comment (beblicarl) [reply src/schema.js:273]: sure, thanks

File: src/routes/posts.js [hunk 8]
```diff
@@ -38,7 +38,7 @@ const fetchPosts = catchAsync( async(req, res) => {
 			})})
 		),
 		metadata: {
-			totalCount: feed.length
+			totalCount: Number(count.count)
```
Comment (jocrah) [inline src/routes/posts.js:41]: kindly handle this in the service before returning it, that's where the main logic is and if we want count to be a number, we should define it there. 

Else wherever we call the same function we would have to do that

File: src/services/posts/fetch-posts.js [hunk 9]
```diff
@@ -31,13 +32,15 @@ const fetchPosts = async({isFlagged, cursor, limit }) => {
...
 ${selectClause} ${fromClause} ${joinClause} 
 ${whereClause} ${orderByClause} ${limitClause};
 `
 	) 
-	
-	return feed.rows
+	const feed = result.rows
+	const value = await db.query(` SELECT COUNT(posts) FROM posts;`)
+	const count = value.rows[0]
+	return {feed , count}
```
Comment (jocrah) [inline src/services/posts/fetch-posts.js:43]: getting quite long and hard to follow what is going on in here, lets try to separate some of the logic into functions

maybe a separate `fetchData` in the same file to deal with the returning the feed itself and `getTotalCount` as a way to get the total count.

and then call them here in the main function

File: src/services/posts/fetch-posts.js [hunk 10]
```diff
@@ -31,13 +32,15 @@ const fetchPosts = async({isFlagged, cursor, limit }) => {
...
-	const feed = await db.query(`
+	const result = await db.query(`
 ${selectClause} ${fromClause} ${joinClause} 
 ${whereClause} ${orderByClause} ${limitClause};
 `
 	) 
-	
-	return feed.rows
+	const feed = result.rows
+	const value = await db.query(` SELECT COUNT(posts) FROM posts;`)
```
Comment (jocrah) [inline src/services/posts/fetch-posts.js:41]: the count should factor in the isFlagged query. 

When the client is fetching all flagged posts, the total count in that case should be the total number of flagged posts

Do remember that, the total count is to help the frontend know whether to make a call for the next set of data or not. So if the number of flagged posts is 3, and the totalCount is 15. The frontend/client will keep making calls thinking there are 15 flagged posts

File: src/routes/posts.js (hunk 8 above)
Comment (beblicarl) [reply src/routes/posts.js]: yes please

File: src/services/posts/fetch-posts.js (hunk 9 above)
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:64]: sure, thanks

File: src/services/posts/fetch-posts.js (hunk 10 above)
Comment (beblicarl) [reply src/services/posts/fetch-posts.js]: this is awesome !

File: src/services/posts/fetch-posts.test.js [hunk 11]
```diff
@@ -18,6 +18,7 @@ describe('Fetch all posts', () => {
 	 })
 	 
 	it('should fetch posts', async () =>{
```
Comment (jocrah) [inline src/services/posts/fetch-posts.test.js:20]: Please remember that, when new code is added, there have to be test cases which capture the changes made to ensure the quality of the code added.  It helps to know code is working well and also helps to know when the code breaks later. A few questions this pull request should be able to answer in the tests are:

1. is the function returning the correct number of posts when a limit is passed?
2. is the function returning the correct data when a cursor is passed?
3. is the correct response (feed and count) being returned from the function?

File: src/services/posts/fetch-posts.js [hunk 12]
```diff
@@ -14,23 +14,54 @@ const fetchPosts = async(isFlagged) => {
...
+	 ${selectClause} ${fromClause} ${whereClause};`)
+	const count = value.rows[0]
+	return Number(count.count)
+}
+const fetchPosts = async(isFlagged, cursor, limit = DEFAULT_LIMIT ) => {
 	
-	return feed.rows
+	const feed = await fetchData(isFlagged, cursor, limit) 
+	const count = await getTotalCount(isFlagged)
+	return {feed , count}
```
Comment (jocrah) [inline src/services/posts/fetch-posts.js:64]: this is cool. Now, as an improvement, we don't have to wait for data to be fetched before counting. 

We can perform both tasks together. Kindly look into [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all). Feel free to look at other resources in relation to it if necessary

File: src/services/posts/fetch-posts.test.js (hunk 11 above)
Comment (beblicarl) [reply src/services/posts/fetch-posts.test.js:20]: yes please, I have taken note of this. Thanks

File: src/services/posts/fetch-posts.js [hunk 13]
```diff
@@ -14,23 +14,54 @@ const fetchPosts = async(isFlagged) => {
...
+		whereClause = (`WHERE posts."flagsCount"
+		 ${isFlagged ? '>': '='} 0`)
+	}
+
+	const value = await db.query(` 
+	 ${selectClause} ${fromClause} ${whereClause};`)
+	const count = value.rows[0]
+	return Number(count.count)
+}
+const fetchPosts = async(isFlagged, cursor, limit = DEFAULT_LIMIT ) => {
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:60]: no reason in particular, i will reverse it.

File: src/services/posts/fetch-posts.js [hunk 14]
```diff
@@ -14,23 +14,54 @@ const fetchPosts = async(isFlagged) => {
...
+	
+		whereClause = (`WHERE posts."flagsCount"
+		 ${isFlagged ? '>': '='} 0`)
+	}
+
+	const value = await db.query(` 
+	 ${selectClause} ${fromClause} ${whereClause};`)
+	const count = value.rows[0]
+	return Number(count.count)
+}
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:59]: sure, i will do that

File: src/services/posts/fetch-posts.js [hunk 15]
```diff
@@ -14,23 +14,54 @@ const fetchPosts = async(isFlagged) => {
...
-	   
-	 if(isFlagged !== undefined){
-		 selectClause+= `, posts."flagsCount"`
-		 
-		whereClause += `
-			WHERE posts."flagsCount" ${isFlagged ? '>': '='} 0`
+	 const limitClause = `LIMIT ${limit}`
+
+
+	// eslint-disable-next-line no-unused-expressions
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/services/posts/fetch-posts.js:22]: it actually does

File: src/services/posts/fetch-posts.test.js [hunk 16]
```diff
@@ -28,9 +29,55 @@ describe('Fetch all posts', () => {
...
 	    INNER JOIN users on posts."userId" = users.id
-		ORDER BY posts."createdAt" DESC;
+		ORDER BY posts."createdAt" DESC LIMIT ${20};
         `)
-		return expect(actualPosts).to.eql(result.rows)	
+        
+		const value = await db.query(`SELECT COUNT(posts) FROM posts;`)
+		const countData = value.rows[0]
+		
+		const {feed, count} = actualPosts
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/services/posts/fetch-posts.test.js:38]: alright sure

File: src/services/posts/fetch-posts.test.js [hunk 17]
```diff
@@ -28,9 +29,55 @@ describe('Fetch all posts', () => {
...
+		return expect(feed, count).to.eql(result.rows, Number(countData.count))
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/services/posts/fetch-posts.test.js:39]: this is very creative!

File: src/services/posts/fetch-posts.test.js [hunk 18]
```diff
@@ -28,9 +29,55 @@ describe('Fetch all posts', () => {
...
+        posts."likesCount",
+        users."firstName", users."lastName", users.email, 
+        users."profilePictureUrl" 
+	    FROM posts
+	    INNER JOIN users on posts."userId" = users.id
+		ORDER BY posts."createdAt" DESC LIMIT ${3};
+        `)
+		
+		const {feed} = actualPosts
+		return expect(feed).to.eql(result.rows)
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/services/posts/fetch-posts.test.js:59]: alright, that's great

File: src/services/posts/fetch-posts.test.js [hunk 19]
```diff
@@ -28,9 +29,55 @@ describe('Fetch all posts', () => {
...
+        posts."likesCount",
+        users."firstName", users."lastName", users.email,
+        users."profilePictureUrl"
+         FROM posts INNER JOIN users on posts."userId" = users.id
+         WHERE  posts."id" > ${cursorValue} 
+		 ORDER BY posts."createdAt" DESC LIMIT ${20};
+        `)
+		
+		const {feed} = actualPosts
+		return expect(feed).to.eql(result.rows)
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/services/posts/fetch-posts.test.js:79]: yeaaah i get you, makes sense, i will do that

File: src/routes/posts.js [hunk 20]
```diff
@@ -23,9 +23,11 @@ const router = express.Router()
 // GET ALL ARTICLES
 
 const fetchPosts = catchAsync( async(req, res) => {
-	const {isFlagged} = req.query
-	const feed = await postService.fetchPosts(isFlagged)
-
+	const {isFlagged, cursor , limit} = req.query
+	const {feed, count} = await postService.fetchPosts({
+		isFlagged, cursor, limit
+	})
```
[the remark this answers is not in the document]
Comment (beblicarl) [reply src/routes/posts.js:29]: oh okay, i get what you mean. I have taken note of this

**Review by jocrah — APPROVED** [review]
great work @beblicarl, Kindly merge at your convenience :rocket:

**Comment by beblicarl** [comment]
> great work @beblicarl, Kindly merge at your convenience 🚀

I am sad it has come to an end, I learned something new everyday lol

**Comment by jocrah** [comment]
> I am sad it has come to an end, I learned something new every day lol.
> Can't wait for new issues

Haha 'sad' is an interesting word choice
